File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> answer = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        helper(nums, target, answer, curr, 0, 0);
        return answer;
    }

    private void helper(int[] nums, int target, 
                        List<List<Integer>> answer,
                        List<Integer> curr, int sum, int index) {
        if (sum < target) {
            for (int i = index; i < nums.length; i++) {
                curr.add(nums[i]);
                helper(nums, target, answer, curr, sum + nums[i], i);
                curr.remove(curr.size() - 1);
            }
        } else if (sum == target) {
            answer.add(new ArrayList<>(curr));
        }
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-0.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        if len(s) < len(t):
            return 0
        dp = {}
        def helper(i, j):
            if j >= len(t):
                return 1
            if i >= len(s):
                return 0
            if (i, j) in dp:
                return dp[(i, j)]
            answer = 0
            if s[i] == t[j]:
                answer += helper(i + 1, j + 1)            
            answer += helper(i + 1, j)
        
            dp[(i, j)] = answer
            return answer
        answer = helper(0, 0)
        print(dp)
        return answer